File: javatests/google/registry/ui/server/registrar/SecuritySettingsTest.java
// Copyright 2016 The Nomulus Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package google.registry.ui.server.registrar;

import static com.google.common.truth.Truth.assertThat;
import static google.registry.config.RegistryConfig.getDefaultRegistrarReferralUrl;
import static google.registry.config.RegistryConfig.getDefaultRegistrarWhoisServer;
import static google.registry.testing.CertificateSamples.SAMPLE_CERT;
import static google.registry.testing.CertificateSamples.SAMPLE_CERT2;
import static google.registry.testing.CertificateSamples.SAMPLE_CERT2_HASH;
import static google.registry.testing.CertificateSamples.SAMPLE_CERT_HASH;
import static google.registry.util.DateTimeUtils.START_OF_TIME;
import static java.util.Arrays.asList;

import com.google.common.collect.ImmutableMap;
import google.registry.model.registrar.Registrar;
import java.util.Map;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.runners.MockitoJUnitRunner;

/**
 * Unit tests for security_settings.js use of {@link RegistrarSettingsAction}.
 *
 * <p>The default read and session validation tests are handled by the
 * superclass.
 */
@RunWith(MockitoJUnitRunner.class)
public class SecuritySettingsTest extends RegistrarSettingsActionTestCase {

  @Test
  public void testPost_updateCert_success() throws Exception {
    Registrar modified = Registrar.loadByClientId(CLIENT_ID).asBuilder()
        .setClientCertificate(SAMPLE_CERT, clock.nowUtc())
        .build();
    Map<String, Object> response = action.handleJsonRequest(ImmutableMap.of(
        "op", "update",
        "args", modified.toJsonMap()));
    // Empty whoisServer and referralUrl fields should be set to defaults by server.
    modified =
        modified
            .asBuilder()
            .setWhoisServer(getDefaultRegistrarWhoisServer())
            .setReferralUrl(getDefaultRegistrarReferralUrl())
            .build();
    assertThat(response).containsEntry("status", "SUCCESS");
    assertThat(response).containsEntry("results", asList(modified.toJsonMap()));
    assertThat(Registrar.loadByClientId(CLIENT_ID)).isEqualTo(modified);
  }

  @Test
  public void testPost_updateCert_failure() throws Exception {
    Map<String, Object> reqJson = Registrar.loadByClientId(CLIENT_ID).toJsonMap();
    reqJson.put("clientCertificate", "BLAH");
    Map<String, Object> response = action.handleJsonRequest(ImmutableMap.of(
        "op", "update",
        "args", reqJson));
    assertThat(response).containsEntry("status", "ERROR");
    assertThat(response).containsEntry("message", "Invalid X.509 PEM certificate");
  }

  @Test
  public void testChangeCertificates() throws Exception {
    Map<String, Object> jsonMap = Registrar.loadByClientId(CLIENT_ID).toJsonMap();
    jsonMap.put("clientCertificate", SAMPLE_CERT);
    jsonMap.put("failoverClientCertificate", null);
    Map<String, Object> response = action.handleJsonRequest(ImmutableMap.of(
        "op", "update", "args", jsonMap));
    assertThat(response).containsEntry("status", "SUCCESS");
    Registrar registrar = Registrar.loadByClientId(CLIENT_ID);
    assertThat(registrar.getClientCertificate()).isEqualTo(SAMPLE_CERT);
    assertThat(registrar.getClientCertificateHash()).isEqualTo(SAMPLE_CERT_HASH);
    assertThat(registrar.getFailoverClientCertificate()).isNull();
    assertThat(registrar.getFailoverClientCertificateHash()).isNull();
  }

  @Test
  public void testChangeFailoverCertificate() throws Exception {
    Map<String, Object> jsonMap = Registrar.loadByClientId(CLIENT_ID).toJsonMap();
    jsonMap.put("failoverClientCertificate", SAMPLE_CERT2);
    Map<String, Object> response = action.handleJsonRequest(ImmutableMap.of(
        "op", "update", "args", jsonMap));
    assertThat(response).containsEntry("status", "SUCCESS");
    Registrar registrar = Registrar.loadByClientId(CLIENT_ID);
    assertThat(registrar.getFailoverClientCertificate()).isEqualTo(SAMPLE_CERT2);
    assertThat(registrar.getFailoverClientCertificateHash()).isEqualTo(SAMPLE_CERT2_HASH);
  }

  @Test
  public void testEmptyOrNullCertificate_doesNotClearOutCurrentOne() throws Exception {
    action.initialRegistrar =
        Registrar.loadByClientId(CLIENT_ID).asBuilder()
            .setClientCertificate(SAMPLE_CERT, START_OF_TIME)
            .setFailoverClientCertificate(SAMPLE_CERT2, START_OF_TIME)
            .build();
    Map<String, Object> jsonMap = action.initialRegistrar.toJsonMap();
    jsonMap.put("clientCertificate", null);
    jsonMap.put("failoverClientCertificate", "");
    Map<String, Object> response = action.handleJsonRequest(ImmutableMap.of(
        "op", "update", "args", jsonMap));
    assertThat(response).containsEntry("status", "SUCCESS");
    Registrar registrar = Registrar.loadByClientId(CLIENT_ID);
    assertThat(registrar.getClientCertificate()).isEqualTo(SAMPLE_CERT);
    assertThat(registrar.getClientCertificateHash()).isEqualTo(SAMPLE_CERT_HASH);
    assertThat(registrar.getFailoverClientCertificate()).isEqualTo(SAMPLE_CERT2);
    assertThat(registrar.getFailoverClientCertificateHash()).isEqualTo(SAMPLE_CERT2_HASH);
  }

  @Test
  public void testToJsonMap_containsCertificate() throws Exception {
    Map<String, Object> jsonMap = Registrar.loadByClientId(CLIENT_ID).asBuilder()
        .setClientCertificate(SAMPLE_CERT2, START_OF_TIME)
        .build()
        .toJsonMap();
    assertThat(jsonMap).containsEntry("clientCertificate", SAMPLE_CERT2);
    assertThat(jsonMap).containsEntry("clientCertificateHash", SAMPLE_CERT2_HASH);
  }

  @Test
  public void testToJsonMap_containsFailoverCertificate() throws Exception {
    Map<String, Object> jsonMap = Registrar.loadByClientId(CLIENT_ID).asBuilder()
        .setFailoverClientCertificate(SAMPLE_CERT2, START_OF_TIME)
        .build()
        .toJsonMap();
    assertThat(jsonMap).containsEntry("failoverClientCertificate", SAMPLE_CERT2);
    assertThat(jsonMap).containsEntry("failoverClientCertificateHash", SAMPLE_CERT2_HASH);
  }
}

File: core/src/test/javascript/google/registry/ui/js/registrar/security_settings_test.js
// Copyright 2017 The Nomulus Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

goog.setTestOnly();

goog.require('goog.dispose');
goog.require('goog.dom');
goog.require('goog.testing.MockControl');
goog.require('goog.testing.PropertyReplacer');
goog.require('goog.testing.net.XhrIo');
goog.require('registry.registrar.ConsoleTestUtil');
goog.require('registry.testing');
goog.require('registry.util');

describe('security settings test', function() {
  const $ = goog.dom.getRequiredElement;
  const stubs = new goog.testing.PropertyReplacer();

  const expectedRegistrar = {
    ipAddressAllowList: [],
    phonePasscode: '12345',
    clientCertificate: null,
    clientCertificateHash: null,
    failoverClientCertificate: null
  };

  const test = {
    testXsrfToken: '༼༎෴ ༎༽',
    testClientId: 'testClientId',
    mockControl: new goog.testing.MockControl()
  };

  beforeEach(function() {
    registry.testing.addToDocument('<div id="test"/>');
    registry.testing.addToDocument('<div class="kd-butterbar"/>');
    registry.registrar.ConsoleTestUtil.renderConsoleMain($('test'), {
      xsrfToken: test.testXsrfToken,
      clientId: test.testClientId,
    });
    stubs.setPath('goog.net.XhrIo', goog.testing.net.XhrIo);
    registry.registrar.ConsoleTestUtil.setup(test);
  });

  afterEach(function() {
    goog.dispose(test.console);
    goog.testing.net.XhrIo.cleanup();
    stubs.reset();
    test.mockControl.$tearDown();
  });

  function testView() {
    registry.registrar.ConsoleTestUtil.visit(test, {
      path: 'security-settings',
      xsrfToken: test.testXsrfToken,
      clientId: test.testClientId
    });
    registry.testing.assertReqMockRsp(
        test.testXsrfToken,
        '/registrar-settings',
        {op: 'read', id: 'testClientId', args: {}},
        {
          status: 'SUCCESS',
          message: 'OK',
          results: [expectedRegistrar]
        });
    expect(registry.util.parseForm('item').phonePasscode).toEqual(expectedRegistrar.phonePasscode);
  }

  it("testView", function() {
    testView();
  });

  it("testEdit", function() {
    testView();

    registry.testing.click($('reg-app-btn-edit'));

    const form = document.forms.namedItem('item');
    form.elements['newIp'].value = '1.1.1.1';
    registry.testing.click($('btn-add-ip'));
    form.elements['newIp'].value = '2.2.2.2';
    registry.testing.click($('btn-add-ip'));

    const exampleCert = $('exampleCert').value;
    const exampleCertHash = '6NKKNBnd2fKFooBINmn3V7L3JOTHh02+2lAqYHdlTgk';
    form.elements['clientCertificate'].value = exampleCert;
    form.elements['failoverClientCertificate'].value = 'bourgeois blues';
    registry.testing.click($('reg-app-btn-save'));

    registry.testing.assertReqMockRsp(
        test.testXsrfToken,
        '/registrar-settings',
        {op: 'update', id: 'testClientId', args: {
          clientCertificate: exampleCert,
          clientCertificateHash: null,
          failoverClientCertificate: 'bourgeois blues',
          ipAddressAllowList: ['1.1.1.1', '2.2.2.2'],
          phonePasscode: expectedRegistrar.phonePasscode,
          readonly: false }},
        {status: 'SUCCESS',
          message: 'OK',
          results: [{}]});
    // XXX: The response above is ignored as the page re-issues a fetch. Should
    //      either provide the real response here and use anyTimes(), or have
    //      resource_component use this directly.

    expectedRegistrar.clientCertificate = exampleCert;
    expectedRegistrar.clientCertificateHash = exampleCertHash;
    expectedRegistrar.failoverClientCertificate = 'bourgeois blues';
    expectedRegistrar.ipAddressAllowList = ['1.1.1.1/32', '2.2.2.2/32'];
    registry.testing.assertReqMockRsp(
        test.testXsrfToken,
        '/registrar-settings',
        {op: 'read', id: 'testClientId', args: {}},
        {status: 'SUCCESS',
          message: 'OK',
          results: [expectedRegistrar]});

    delete expectedRegistrar['clientCertificateHash'];
    registry.testing.assertObjectEqualsPretty(
        expectedRegistrar, registry.util.parseForm('item'));
  });
});
